### ISSUE TITLE
git master fails build/tests on i386 and other 32-bit architectures

### ISSUE THREAD

**Issue by smcv** [flatpak/flatpak#3499]
## Linux distribution and version

Debian unstable

## Flatpak version

git master as of yesterday (commit 05eba2b3) plus #3488 and #3489 applied as patches

## Description of the problem

```
common/flatpak-dir.c: In function ‘flatpak_dir_pull_untrusted_local’:
common/flatpak-dir.c:5468:115: error: format ‘%ld’ expects argument of type ‘long int’, but argument 6 has type ‘guint64’ {aka ‘long long unsigned int’} [-Werror=format=]
 5468 |         return flatpak_fail_error (error, FLATPAK_ERROR_DOWNGRADE, "Not allowed to downgrade %s (old_commit: %s/%ld new_commit: %s/%ld ",
      |                                                                                                                 ~~^
      |                                                                                                                   |
      |                                                                                                                   long int
      |                                                                                                                 %lld
 5469 |                                    ref, current_checksum, old_timestamp, checksum, new_timestamp);
      |                                                           ~~~~~~~~~~~~~                                            
      |                                                           |
      |                                                           guint64 {aka long long unsigned int}
common/flatpak-dir.c:5468:134: error: format ‘%ld’ expects argument of type ‘long int’, but argument 8 has type ‘guint64’ {aka ‘long long unsigned int’} [-Werror=format=]
 5468 |         return flatpak_fail_error (error, FLATPAK_ERROR_DOWNGRADE, "Not allowed to downgrade %s (old_commit: %s/%ld new_commit: %s/%ld ",
      |                                                                                                                                    ~~^
      |                                                                                                                                      |
      |                                                                                                                                      long int
      |                                                                                                                                    %lld
 5469 |                                    ref, current_checksum, old_timestamp, checksum, new_timestamp);
      |                                                                                    ~~~~~~~~~~~~~                                      
      |                                                                                    |
      |                                                                                    guint64 {aka long long unsigned int}
```

and also a lot of warnings like this:

```
In file included from /usr/include/string.h:495,
                 from common/flatpak-dir.c:29:
In function ‘memcpy’,
    inlined from ‘var_ref_read_unaligned_le’ at common/flatpak-variant-impl-private.h:44:9,
    inlined from ‘var_deploy_data_get_metadata’ at common/flatpak-variant-impl-private.h:2623:20,
    inlined from ‘flatpak_deploy_data_get_localed_string’ at common/flatpak-dir.c:2554:29:
/usr/include/i386-linux-gnu/bits/string_fortified.h:34:10: warning: ‘__builtin___memcpy_chk’ forming offset [5, 8] is out of the bounds [0, 4] of object ‘tmpvalue’ with type ‘union <anonymous>’ [-Warray-bounds]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from common/flatpak-dir.c:57:
common/flatpak-dir.c: In function ‘flatpak_deploy_data_get_localed_string’:
common/flatpak-variant-impl-private.h:37:5: note: ‘tmpvalue’ declared here
   37 |   } tmpvalue;
      |     ^~~~~~~~
```

## Steps to reproduce

Build Flatpak on i386 or another ILP32 architecture with `-Werror=format-security` (which is done by default for most official Debian packages), e.g. https://buildd.debian.org/status/fetch.php?pkg=flatpak&arch=i386&ver=1.7.0%7Egit20200325-1&stamp=1585171309&raw=0, https://buildd.debian.org/status/fetch.php?pkg=flatpak&arch=armhf&ver=1.7.0%7Egit20200325-1&stamp=1585171074&raw=0

**Comment by smcv** [comment]
#3500 fixes the format string issues.

I'm not so sure about the variant-schema-compiler issue. @alexlarsson, do you have any more minimal test-cases for it that I could try on various architectures? I tried this:

```
diff --git a/variant-schema-compiler b/variant-schema-compiler
index 763c1ab..ffd6c05 100755
--- a/variant-schema-compiler
+++ b/variant-schema-compiler
@@ -140,13 +140,13 @@ G_GNUC_CONST static inline guint
 #endif
 }}
 
-G_GNUC_PURE static inline gsize
+G_GNUC_PURE static inline guint64
 {prefix_}ref_read_unaligned_le (guchar *bytes, guint   size)
 {{
   union
   {{
-    guchar bytes[GLIB_SIZEOF_SIZE_T];
-    gsize integer;
+    guchar bytes[8];
+    guint64 integer;
   }} tmpvalue;
 
   tmpvalue.integer = 0;
@@ -166,7 +166,7 @@ G_GNUC_PURE static inline gsize
         memcpy (&tmpvalue.bytes, bytes, 1);
     }}
 
-  return GSIZE_FROM_LE (tmpvalue.integer);
+  return GUINT64_FROM_LE (tmpvalue.integer);
 }}
 
 static inline void
```

which silenced the warnings, but the tests failed with a segmentation fault.

I'm also seeing lots of these errors on ppc64el (which is little-endian LP64, so I would have expected it to behave a lot like x86_64 and aarch64, which succeeded) and on ppc64 (big-endian LP64):

```
ERROR: tests/test-completion.sh - Bail out! flatpak:ERROR:common/flatpak-variant-impl-private.h:2190:var_sparse_cache_entry_get_key: assertion failed: (end < v.size && base[end-1] == 0)
```

and a floating point exception on s390x (big-endian LP64), which makes me suspect that the code produced by the variant schema compiler is not particularly portable.

**Comment by smcv** [comment]
Build logs for all official and unofficial Debian architectures: https://buildd.debian.org/status/package.php?p=flatpak&suite=experimental

(Needs-Build on mips(64)el just means they haven't tried it yet, and a bunch of unofficial architectures say BD-Uninstallable because they don't have the libseccomp build-dependency; the interesting ones are the Build-Attempted status.)

**Comment by alexlarsson** [comment]
@smcv if you check out https://gitlab.gnome.org/alexl/variant-schema-compiler/ there are some smaller testcases. For example you can try "make performance", which creates some GVariants and tries to parse them.

You could tweak the makefile to pass --internal-validation to the generator which might help

**Comment by smcv** [comment]
Thanks, I've opened https://gitlab.gnome.org/alexl/variant-schema-compiler/-/merge_requests/1 https://gitlab.gnome.org/alexl/variant-schema-compiler/-/merge_requests/2 and https://gitlab.gnome.org/alexl/variant-schema-compiler/-/merge_requests/3 with some steps towards this.

**Comment by smcv** [comment]
```
GVariant performance: 25.7 kiloiterations per second
**
ERROR:performance.h:888:performance_container_get_tuple: assertion failed: (offset + 4 < v.size)
Bail out! ERROR:performance.h:888:performance_container_get_tuple: assertion failed: (offset + 4 < v.size)
```

**Comment by smcv** [comment]
Looks like at least part of the problem is that the `__OPTIMIZE__` code path in `performance_ref_get_offset_size()` doesn't produce the same result as the straightforward code path, at least on i386. When I hack it to try both calculations and compare them, I get:

```
** (process:347205): ERROR **: 13:39:47.630: disagrees on size=48: optimized 8, unoptimized 1
```

**Comment by smcv** [comment]
https://gitlab.gnome.org/alexl/variant-schema-compiler/-/merge_requests/4 seems to be enough to make flatpak pass tests on i386 again. I haven't tried a big-endian platform like s390x or ppc64 yet.

**Comment by alexlarsson** [comment]
I updated the submodule in #3501 
Would love to hear if it also fixes the other arches.

**Comment by smcv** [comment]
I have access to various Debian machines, including one specimen of each official "release architecture", so I'll try latest git on those.

**Comment by smcv** [comment]
s390x (LP64 BE) is also fixed by this, to the extent that I can test it (bwrap doesn't work inside a schroot environment).

mips (ILP32 BE) I'm waiting on a build.

ppc64el (LP64 LE) is not fixed by this, I'll open a separate issue.

**Comment by smcv** [comment]
mips also seems fine.
